File: backend/app/api/ftp.py
"""FTP Server API endpoints for managing FTP services and users."""

from flask import Blueprint, request, jsonify

from ..middleware.rbac import admin_required, viewer_required
from ..services.ftp_service import FTPService

ftp_bp = Blueprint('ftp', __name__)


@ftp_bp.route('/status', methods=['GET'])
@viewer_required
def get_status():
    """Get FTP server status."""
    result = FTPService.get_status()
    return jsonify(result), 200


@ftp_bp.route('/service/<action>', methods=['POST'])
@admin_required
def control_service(action):
    """Start, stop, restart, or reload FTP service."""
    if action not in ['start', 'stop', 'restart', 'reload']:
        return jsonify({
            'success': False,
            'error': 'Invalid action. Use: start, stop, restart, or reload'
        }), 400

    data = request.get_json() or {}
    service = data.get('service')  # Optional: specify vsftpd or proftpd

    result = FTPService.control_service(service, action)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/config', methods=['GET'])
@viewer_required
def get_config():
    """Get FTP server configuration."""
    service = request.args.get('service')  # Optional
    result = FTPService.get_config(service)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/config', methods=['POST'])
@admin_required
def update_config():
    """Update FTP server configuration."""
    data = request.get_json()

    if not data:
        return jsonify({'success': False, 'error': 'Request body is required'}), 400

    service = data.get('service')
    config_content = data.get('config')

    if not config_content:
        return jsonify({'success': False, 'error': 'Configuration content is required'}), 400

    result = FTPService.update_config(service, config_content)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/users', methods=['GET'])
@viewer_required
def list_users():
    """List FTP users."""
    result = FTPService.list_users()

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/users', methods=['POST'])
@admin_required
def create_user():
    """Create a new FTP user."""
    data = request.get_json()

    if not data:
        return jsonify({'success': False, 'error': 'Request body is required'}), 400

    username = data.get('username')
    password = data.get('password')  # Optional, can be auto-generated
    home_dir = data.get('home_dir')  # Optional

    if not username:
        return jsonify({'success': False, 'error': 'Username is required'}), 400

    # Validate username
    if not username.isalnum() and '_' not in username:
        return jsonify({
            'success': False,
            'error': 'Username must contain only alphanumeric characters and underscores'
        }), 400

    result = FTPService.create_user(username, password, home_dir)

    if result.get('success'):
        return jsonify(result), 201
    return jsonify(result), 400


@ftp_bp.route('/users/<username>', methods=['DELETE'])
@admin_required
def delete_user(username):
    """Delete an FTP user."""
    delete_home = request.args.get('delete_home', 'false').lower() == 'true'

    result = FTPService.delete_user(username, delete_home=delete_home)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/users/<username>/password', methods=['POST'])
@admin_required
def change_password(username):
    """Change FTP user password."""
    data = request.get_json() or {}
    new_password = data.get('password')  # Optional, can be auto-generated

    result = FTPService.change_password(username, new_password)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/users/<username>/toggle', methods=['POST'])
@admin_required
def toggle_user(username):
    """Enable or disable an FTP user."""
    data = request.get_json() or {}
    enabled = data.get('enabled', True)

    result = FTPService.toggle_user(username, enabled)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/connections', methods=['GET'])
@viewer_required
def get_connections():
    """Get active FTP connections."""
    result = FTPService.get_connections()

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/connections/<int:pid>', methods=['DELETE'])
@admin_required
def disconnect_user(pid):
    """Disconnect an active FTP session."""
    result = FTPService.disconnect_session(pid)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/logs', methods=['GET'])
@viewer_required
def get_logs():
    """Get FTP server logs."""
    lines = request.args.get('lines', 100, type=int)

    result = FTPService.get_logs(lines)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/install', methods=['POST'])
@admin_required
def install_ftp():
    """Install FTP server (vsftpd or proftpd)."""
    data = request.get_json() or {}
    service = data.get('service', 'vsftpd')

    if service not in ['vsftpd', 'proftpd']:
        return jsonify({
            'success': False,
            'error': 'Invalid service. Use: vsftpd or proftpd'
        }), 400

    result = FTPService.install_ftp_server(service)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400


@ftp_bp.route('/test', methods=['POST'])
@admin_required
def test_connection():
    """Test FTP connection."""
    data = request.get_json() or {}
    host = data.get('host', 'localhost')
    port = data.get('port', 21)
    username = data.get('username')
    password = data.get('password')

    result = FTPService.test_connection(host, port, username, password)

    if result.get('success'):
        return jsonify(result), 200
    return jsonify(result), 400

File: backend-rs/crates/sk-server/src/routes/ftp.rs
use crate::error::{ApiError, ApiResult};
use crate::extract::AuthUser;
use crate::state::SharedState;
use axum::extract::{Path, Query, State};
use axum::routing::{delete, get, post};
use axum::{Json, Router};
use serde::Deserialize;
use serde_json::{json, Value};

fn internal(e: anyhow::Error) -> ApiError {
    e.into()
}
fn body(b: Option<Json<Value>>) -> Value {
    match b.map(|x| x.0).unwrap_or_else(|| json!({})) {
        Value::String(s) => serde_json::from_str(&s).unwrap_or_else(|_| json!({"value": s})),
        v => v,
    }
}

pub fn router() -> Router<SharedState> {
    Router::new()
        .route("/status", get(status))
        .route("/config", get(config).post(set_config))
        .route("/users", get(users).post(create_user))
        .route("/users/{username}", delete(delete_user))
        .route("/users/{username}/password", post(change_password))
        .route("/users/{username}/toggle", post(toggle_user))
        .route("/connections", get(connections))
        .route("/connections/{pid}", delete(disconnect))
        .route("/logs", get(logs))
        .route("/install", post(install))
        .route("/service/{action}", post(service))
        .route("/test", post(test))
}

#[derive(Deserialize)]
struct Q {
    service: Option<String>,
    lines: Option<i64>,
    delete_home: Option<bool>,
}
async fn status(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::status(&s.db).await.map_err(internal)?))
}
async fn config(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Query(q): Query<Q>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::config(&s.db, q.service.as_deref())
            .await
            .map_err(internal)?,
    ))
}
async fn set_config(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::set_config(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn users(State(s): State<SharedState>, AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::users(&s.db).await.map_err(internal)?))
}
async fn create_user(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::create_user(&s.db, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn delete_user(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(username): Path<String>,
    Query(q): Query<Q>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::delete_user(&s.db, &username, q.delete_home.unwrap_or(false))
            .await
            .map_err(internal)?,
    ))
}
async fn change_password(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(username): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::change_password(&s.db, &username, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn toggle_user(
    State(s): State<SharedState>,
    AuthUser(_): AuthUser,
    Path(username): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::toggle_user(&s.db, &username, &body(b))
            .await
            .map_err(internal)?,
    ))
}
async fn connections(AuthUser(_): AuthUser) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::connections().await.map_err(internal)?))
}
async fn disconnect(AuthUser(_): AuthUser, Path(pid): Path<String>) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::disconnect(&pid).await.map_err(internal)?))
}
async fn logs(AuthUser(_): AuthUser, Query(q): Query<Q>) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::logs(q.lines.unwrap_or(100))
            .await
            .map_err(internal)?,
    ))
}
async fn install(AuthUser(_): AuthUser, b: Option<Json<Value>>) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::install(&body(b)).await.map_err(internal)?))
}
async fn service(
    AuthUser(_): AuthUser,
    Path(action): Path<String>,
    b: Option<Json<Value>>,
) -> ApiResult<Json<Value>> {
    Ok(Json(
        sk_ftp::service(&action, &body(b)).await.map_err(internal)?,
    ))
}
async fn test(AuthUser(_): AuthUser, b: Option<Json<Value>>) -> ApiResult<Json<Value>> {
    Ok(Json(sk_ftp::test(&body(b)).await.map_err(internal)?))
}
